### PR TITLE
save attribute POST endpoint

### PR DESCRIPTION
**Pull request by SanaBob**
Json schema for post request body
`
{
    "id": Optional<Integer>,
    "name": Required<String>,
    "question": Required<String>,
    "value_type": Required<ENUM[“String”, “Number”, “TeamTemplateSlug”]>,
    "max_selections": Required<Integer>,
    "team_set_template": Required<null>,
    "course": Required<Integer>,
    "options": Required[
      {
        "id": Optional<Integer>,
        "value": Required<String>,
        "label": Required<String>
      },
    ]
}
`

Return
`
{ "data": {
  "id": <Integer>,
  "options": [
    {
      "id": <Integer>,
      "created_at": <Datetime>,
      "updated_at": <Datetime>,
      "label": <String>,
      "value": <String>,
      "attribute": <Integer>
    }
  ],
  "created_at": <Datetime>,
  "updated_at": <Datetime>,
  "name": <String>,
  "question": <String>,
  "value_type": <String>,
  "max_selections": <Integer>,
  "team_set_template": <Null>,
  "course": <Integer>
}
}
`

### CODE REVIEW

File: app/tests.py [hunk 1]
```diff
@@ -1,3 +1,244 @@
...
+    factory_request = factory.post(
+        "/api/v1/attributes/save_attribute/",
+        content_type="application/json",
+        data=json.dumps(data),
+    )
+    request = Request(factory_request, parsers=[JSONParser()])
+    return request
+
+
+if __name__ == "__main__":
```
Comment (Opeyem1a) [inline app/tests.py:243]: why is this here?

**Review by Opeyem1a — CHANGES_REQUESTED** [review]

File: app/tests.py (hunk 1 above)
Comment (SanaBob) [reply app/tests.py:243]: I believe this is required to run unittest test suite

File: app/views/attribute.py [hunk 2]
```diff
@@ -14,6 +18,65 @@ class AttributeViewSet(viewsets.ModelViewSet):
...
+        if not attribute_option_serializer.is_valid():
+            raise ValidationError(attribute_option_serializer.errors)
+        attribute = get_object_or_404(Attribute, pk=attribute_id)
+
+        attribute_option_id = attribute_option_data.get("id")
+        if attribute_option_id:
+            attribute_option = get_object_or_404(
+                AttributeOption, pk=attribute_option_id
+            )
+            AttributeOption.objects.filter(pk=attribute_option_id).update(
```
Comment (Opeyem1a) [inline app/views/attribute.py:36]:
```suggestion
            attribute_option.update(
```

Don't you already have the object instance that you want to update?

File: app/views/attribute.py [hunk 3]
```diff
@@ -14,6 +18,65 @@ class AttributeViewSet(viewsets.ModelViewSet):
...
+            )
+        else:
+            attribute_option = AttributeOption.objects.create(
+                **{**attribute_option_data, "attribute": attribute}
+            )
+
+        return attribute_option
+
+    @action(detail=False, methods=["post"])
+    def save_attribute(self, request):
```
Comment (Opeyem1a) [inline app/views/attribute.py:47]: this function likely needs to be in an atomic transaction - See `@transaction.atomic`

**Review by Opeyem1a — CHANGES_REQUESTED** [review]

File: app/views/attribute.py (hunk 2 above)
Comment (SanaBob) [reply app/views/attribute.py:36]: https://stackoverflow.com/questions/67250135/why-is-it-not-possible-to-update-a-django-model-instance-like-this-instanceu
I believe there's no update function for django object instance unfortunately so we cannot do .update on attribute_option

**Review by Opeyem1a — APPROVED** [review]

**Pull request merged**